### ISSUE TITLE
confirm_publish defaults to False and insufficiently documented

### ISSUE THREAD

**Issue by arnonmoscona** [celery/celery#5410]
# Description
When using RabbitMQ, the `confirm_publish` argument to the transport layer defaults to `False` (this is the default of py-amqp). When RabbitMQ is stressed (hit the memory high water mark, or out of disk space) it would start dropping messages. When `confirm_publish` is `False` this results in silent message loss when invoking Celery tasks (publish side). Calls to `apply_async()` or `delay()` return immediately and without error even though the message was never published.

If `BROKER_TRANSPORT_OPTIONS` includes `'confirm_publish': True` then the calls to to `apply_async()` or `delay()` will block until the message publish is confirmed by RabbitMQ, and if RabbitMQ remains stressed then eventually you get and exception that you can handle or would at least show in your logs.

None of this is documented. The only example given for `BROKER_TRANSPORT_OPTIONS` is `broker_transport_options = {'visibility_timeout': 18000}  # 5 hours`. And that too, is nowhere near the documentation for `apply_async()`.

# Suggestions
* Provide an example of turning on `confirm_publish` in the documentation of `BROKER_TRANSPORT_OPTIONS`, with an accompanying explanation on the effects of this on RabbitMQ and publishing reliability.
* Provide equivalent documentation for other transports
* Make note on publishing reliability in the documentation of `apply_async()` and refer to the documentation of `BROKER_TRANSPORT_OPTIONS` there.

**Comment by michael-wilg-8** [comment]
Agree. I think it is critical for at-least-once delivery.

**Comment by auvipy** [comment]
Would love to see a PR for the docs improvement

**Comment by pdonadeo** [comment]
I read a lot of pain in this issue, the same pain I'm experiencing since a week. Thanks @arnonmoscona, you solved my issue.

It's a shame to see such a good and mainstream project having such a poor documentation.

**Comment by thedrow** [comment]
> It's a shame to see such a good and mainstream project having such a poor documentation.

I completely agree. We'd accept a PR to fix this issue and further documentation issues you may find.
Contributions are always welcome.

**Comment by meensu23** [comment]
@arnonmoscona Thanks! 
Is there an accompanying timeout or confirm_timeout ? Can confirm_publish infinitely hang?

**Comment by jseadragon** [comment]
Ran into this recently and was really shocked this was the default. I am going to make time to submit a PR to update the documentation.

**Comment by auvipy** [comment]
please move forward with a PR
